### PR TITLE
adding an WDC2704M display to the Net-IO addon board via I2C

### PR DESCRIPTION
**Pull request by cyc1ingsir**
A Wintek WD-C2704M lcd display (which was in store at Pollin's a few years ago) can be attached via an I2C expander.
It does include two HD44780 controller to drive it's four lines a 27 characters.
Since it doesn't has a backlight, it can be wired to the Net-IO addon board omitting the backlight circuitry on the addon board and using the pin for the second EN instead.
This patch does include the pin configuration for the addon board only. After defining the pin config, it may be used on a different board too.
My starting point for connecting the display was this one:
http://www.cc-zwei.de/wiki/index.php?title=ATM18::Forumsprojekt::Wintek_WD-C2704M_an_I2C

### CODE REVIEW

File: pinning/hardware/netio_addon.m4 [hunk 1]
```diff
@@ -53,8 +53,11 @@ ifdef(`conf_DCF77', `dnl
   pin(DCF1_PON, PA1, OUTPUT)
 ')dnl
 
-ifelse(value_HD44780_CONNECTION,`HD44780_I2CSUPPORT',`dnl
+ifelse(value_HD44780_TYPE,`HD44780_WDC2704',`dnl
```
Comment (michaelb42) [inline pinning/hardware/netio_addon.m4]: HD44780_WDC2704 and I2CSUPPORT should be one of many valid combinations only - defining the mapping that way results in wrong mapping if the display is connected directly.

File: hardware/lcd/hd44780_i2c.c [hunk 2]
```diff
@@ -97,12 +118,13 @@ output_nibble(uint8_t rs, uint8_t nibble, uint8_t en)
   if (rs)
     lcd_data |= _BV(HD44780_PCF8574x_RS);
 
+#if HD44780_TYPE != HD44780_WDC2704
```
Comment (michaelb42) [inline hardware/lcd/hd44780_i2c.c]: Really disable backlight support for every connection type of this display when it is only required for i2c?

File: hardware/lcd/hd44780_i2c.c [hunk 3]
```diff
@@ -133,6 +155,7 @@ input_nibble(uint8_t rs, uint8_t en)
 void
 hd44780_backlight(uint8_t state)
 {
+#if HD44780_TYPE != HD44780_WDC2704
```
Comment (michaelb42) [inline hardware/lcd/hd44780_i2c.c]: see above.

File: hardware/lcd/hd44780_i2c.c [hunk 4]
```diff
@@ -149,10 +173,12 @@ hd44780_hw_init(void)
 {
   /* alle ausgaenge auf 0 setzen */
   lcd_data = 0;
+#if HD44780_TYPE != HD44780_WDC2704
```
Comment (michaelb42) [inline hardware/lcd/hd44780_i2c.c]: see above.

File: hardware/lcd/config.in [hunk 5]
```diff
@@ -31,6 +31,13 @@ mainmenu_option next_comment
       define_bool I2C_MASTER_SUPPORT y
       define_bool I2C_PCF8574X_SUPPORT y
     fi
+
+    if [ "$HD44780_I2CSUPPORT" = "y" ]; then
+      if [ "$HD44780_TYPE" = "HD44780_WDC2704" ]; then
+        define_string CONF_HD44780_TYPE "HD44780_WDC2704"
```
Comment (michaelb42) [inline hardware/lcd/config.in]: This is only one of many possible CONF_HD44780_TYPE.. Please take care of the others by providing this one for the supported combination. And please do not mix the differnt HD44780-style display-types with their connection.

**Comment by cyc1ingsir** [comment]
Hi Michael,
thanks for reviewing the PR so swiftly.
As far as I know, the WD-C2704M has no backlight build into it at all.
Looking into archive provided for download at the Pollin page (see  http://old.ethersex.de/index.php/WINTEK_WD-C2704M-1HNN for a reference) one can find a datasheet (pdf named WD-C2704M-1HNN.pdf) which supports this theory. So do all other references to this display when searching the net.
IMHO the #ifdef backlight exclusion for all WD2704 is a valid one.
On the other hand: Shouldn't all code dealing with the backlight control depend on  HD44780_BACKLIGHT_SUPPORT
But this wouldn't work for hd44780_i2c.c since it depends on "$HD44780_DIREKT" = "y" (in config.in)

Defining the CONF_HD44780_TYPE in the config.in might be a bit of a hack. My first approach was to test for "HD44780_TYPE" within the file://pinning/hardware/netio_addon.m4 which doesn't work as types without a leading CONF_ are not evaulated.
In my opinion, the cleaner solution would be to change the line
'Original'                HD44780_TYPE
to
'Original'                CONF_HD44780_TYPE
and change all 11 references to HD44780_TYPE appropriately. 
However, I'm concerned of the implications involved.

In the end I do see two ways the WD-C2704M might be connected:
ethersex-board -> I2C - expander -> WD-C2704M
ethersex-board -> WD-C2704M (HD44780_DIREKT)

For the latter, it's important that HD44780_MULTIENSUPPORT support is defined 
which is then evaluated within the hd44780.c
(that's where I got the idea for extending the hd44780_i2c.c from actually).
The wiring is specific to the individual board.

What do you suggest?

Cheers
cyclingsir

**Comment by michaelb42** [comment]
Simply said: separate the display/controller type from its connection type and do not confuse these definitions. These are entirely different settings although for some displays there might be only one possible connection variant.

The controller type define is already there. For the m4 magic all config symbols CONF_xyz are rewritten to value_xyz (and abc_SUPPORT to conf_abc :grin: ). Imo one might replace all occurences of HD44780_TYPE with CONF_HD44780_TYPE - but please make sure that this change is extensively tested.

For the two possible connection types you've outlined above simply put an ifelse inside of the scope of ```ifelse(value_HD44780_CONNECTION,`HD44780_I2CSUPPORT',`dnl```

something like

```
ifelse(value_HD44780_CONNECTION,`HD44780_I2CSUPPORT',`dnl
ifelse (value_HD44780_TYPE,`HD44780_WDC2704',`dnl
dnl HD44780_PCF8574x_MULTI_MAPPING(ADR,RS,RW,EN1,EN2,DB4,DB5,DB6,DB7)
HD44780_PCF8574x_MULTI_MAPPING(0x20, 4, 5, 6, 7, 0, 1, 2, 3)',`dnl
dnl HD44780_PCF8574x_MAPPING(ADR,RS,RW,EN,DB4,DB5,DB6,DB7,BL)
HD44780_PCF8574x_MAPPING(0x20, 4, 5, 6, 0, 1, 2, 3, 7)')')dnl
``` 

Think of it as the beginning of a switch-statement for different display types connected via I2C.

Regarding the backlight and the multiple enable support you should definitly change lcd/config.in to reflect the valid choices. The existing code is simply wrong in assuming that all i2c displays have a backlight and never use multiple enable pins. And imo it's really a no-no that a define in header.m4 re-defines / has to re-define defines introduced in config.in.

**Comment by eku** [comment]
Will you rework your code?

**Comment by cyc1ingsir** [comment]
Finally, I pushed the changes made.
There wasn't any response on the mailing list about testing the separation.
However I can confirm that the changes made do work on a WDC2704M connected via I2C on a NetIO with a addon board.
I have further tested it with another  HD44780 based display directly connected to an etherrape  board. The LCD does have a backlight so I could test switching on an off the backlight via ecmd as well.
I couldn't make the latter tests earlier than today though, hence the delay.

**Comment by eku** [comment]
@cyc1ingsir  could you please rework the PR and remove the EMS commits.

**Pull request closed without merging**